src/Http/Http.Features/src/IHttpRequestCanHaveBodyFeature.cs

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Http.Features
+{
+    /// <summary>
+    /// Used to indicate if the request may have a body.
+    /// </summary>
+    public interface IHttpRequestCanHaveBodyFeature
+    {
+        /// <summary>
+        /// Indicates if the request may have a body. Returns null if the server has not implemented the feature.
+        /// </summary>
+        /// <remarks>
+        /// This returns true when:
+        /// - It's an HTTP/1.x request with a non-zero Content-Length or a 'Transfer-Encoding: chunked' header.
+        /// - It's an HTTP/2 request that did not set the END_STREAM flag on the initial headers frame.
+        /// The final request body length may still be zero for the chunked or HTTP/2 scenarios.
+        ///
+        /// This returns false when:
+        /// - It's an HTTP/1.x request with no Content-Length or 'Transfer-Encoding: chunked' header, or the Content-Length is 0.
+        /// - It's an HTTP/2 request that set END_STREAM on the initial headers frame.
+        /// When false, the request body should never return data.
+        ///
+        /// This returns null if the underlying server has not implemented the feature.
+        /// </remarks>
+        bool? CanHaveBody { get; }
```

Comment:
Should this be HasBody?

Reply:
That's what I thought initially, but there's a case where that get's misleading. If you're doing streaming (chunked or H2) all we know up front is that there might be a body. The end result could still be zero bytes of body data received. Would it be confusing to see HasBody, but then ReadAsync returns 0?

Reply:
With H2, when would this return false? End stream flag is on the request HEADERS frame?

With H2, when would this return true but have a zero body? No end stream flag on the HEADERS frame, then a zero length DATA frame with end stream flag?

Would it still return true if both frames were received simultaneously? (I'm guessing it would)

I wouldn't find HasBody with a zero length body confusing.

Reply:
> With H2, when would this return false? End stream flag is on the request HEADERS frame?

Correct.

> With H2, when would this return true but have a zero body? No end stream flag on the HEADERS frame, then a zero length DATA frame with end stream flag?

Correct, or end stream on a trailers frame without data in between.
 
> Would it still return true if both frames were received simultaneously? (I'm guessing it would)

Yes, frames are processed one at a time and the request is dispatched as soon as the headers are available.